### ISSUE TITLE
Custom Leaspy exceptions, PyTorch >= 1.7 compat, type hints, cleaner doc & cleaner CI - [merged]

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#176]
In GitLab by @emaheux on Sep 23, 2021, 16:41

_Merges feat_leaspy_exceptions -> dev_

# Changes

- [FEAT] Custom Leaspy exceptions raised in code: cf. `leaspy.exceptions` in documentation
- [COMPAT] PyTorch >=1.7 is now supported, as well as Python 3.9 (I tested torch 1.7.1, 1.8.1 & last one: 1.9.1)
- [CODE] Broad use of type annotations in Leaspy package
- [COMPAT] As a consequence support of Python 3.6 is dropped
- [CI] Cleaner CI without need for tox, with one-shot conda envs for Python 3.7, 3.8, 3.9
- [DOC] cleaned
- \+ Premises to clean the `loss` parameter (cf. !53): to be continued (either in this MR as a new commit or anoter?)  

# What to start with

Many many changes due to transversal changes made (type hints + clean doc + custom exceptions).
Maybe the best would be to check changes on files, commit per commit?

The code is iso-functional apart for last commit that implement preliminary work for properly handling loss setting in Leaspy.

# Related issues

#40

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Sep 23, 2021, 16:41

requested review from @Symcies

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Sep 23, 2021, 16:44

changed target branch from `master` to `dev`

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Sep 24, 2021, 14:02

added 1 commit

<ul><li>c15af178 - [FIX] loss should only belong to model, add a check in scipy minimize with an...</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/54/diffs?diff_id=250830072&start_sha=0986c20b80785f5c8afb19e333927ecf77e0f5c0)

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 16:59

Commented on [leaspy/exceptions.py line 25](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-6d94288d507c946892fa5c04b537c678R25)



I wonder if it shouldn't be elsewhere (a read-me for future devs?) or it is too much a pain?

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 17:34

Commented on [leaspy/io/data/data.py line 160](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-7bf8e54555d1e354ec3979f66cfc3711R160)



Shouldn't we assert that the cofactors are present in the individual data, meaning should we raise a clear exception rather than something that will say `GENDER not in keys()`?

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 18:05

Commented on [leaspy/algo/abstract_algo.py line 23](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-23ebe00b397d9ab4e4ababa3805b3079L23)



Is the usage of `:mod:`, `:class:`, ... documented somewhere? I wonder if we should not write (if it is not yet) the correct documentation format that we expect from next devs.

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 18:12

Commented on [leaspy/io/logs/visualization/plotter.py line 47](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-c9c3894d7799e493bdc4bbb8b3bdefb2R47)



I agree that while being identical here, it can lead to unexpected results if, for example, someone writes a `logistic-new-DL-awesome-model` that has nothing to do with what previous devs wrote in `logistic` and `logistic_parallel`.
I think it was more of a "garde fou" to list the "allowed models".

PS : This is a minor comment as I'm not sure many users use the `plotter` so consider it more from a "metaphysical" / "conventional" point of view.

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 18:20

Commented on [leaspy/io/outputs/result.py line 173](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..274050d8b10560b63132edcc7c976d7edeeff8f2#diff-767d0ac58b5e23c2fb2d7c0875ced777R173)



PEP8 recommends to 

> Don't use spaces around the = sign when used to indicate a keyword argument, or when used to indicate a default value for an unannotated function parameter

Which means to use `cofactors=None` rather than `cofactors = None`. 

However, they do recommand to use one if there is an argumentation annotation. So I think it should be : 

`save_individual_parameters_csv(self, path: str, idx: List[IDType] = None, cofactors=None, **args)`. I know it is very minor but PyCharm can be a pain with PEP8 recommandations (which leads to error detection).

[Source](https://www.python.org/dev/peps/pep-0008/#other-recommendations)

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 18:21

Commented on [leaspy/io/outputs/result.py line 220](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..274050d8b10560b63132edcc7c976d7edeeff8f2#diff-767d0ac58b5e23c2fb2d7c0875ced777R220)



Same for `human_readable=None` and elsewhere but I won't indicated it as it is quite minor and could be solve along the flow in other MRs.

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 18:24

Commented on [leaspy/io/settings/algorithm_settings.py line 403](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-181e761aa9fab92bbfc7f4ac86823910R403)



Nice ! I always wondered what is the advantage of the `format` method !

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 18:37

Commented on [leaspy/models/abstract_model.py line 16](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-196e847e3c942e778868226a64191c31R16)



I'm not 100% sure to understand this `TYPE_CHECKING`. Esp. as it is part of the `typing` package which sets it to 

```
# Constant that's True when type checking, but False here.
TYPE_CHECKING = False
```

(line 1480 of `typing.py`)

If I understand correctly, it set to True, it will check the types that you detail elsewhere in the different method signatures and outputs, right? If so, when / in which context is it set to True?

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 18:45

Commented on [CHANGELOG.md line 14](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR14)



Maybe you should explain these changes, both from a user and dev perspective, at a Leaspy(pe) meeting?

I think, esp. for future devs, it could be useful to know some good practices & tricks that you are using in this MR, e.g. the exceptions, the annotations, the `suffixed_method`, how to prevent warning to pop up in the tests, some PEP8 conventions, etc. 

At least I would have more insights from you and I believe other can benefit from this knowledge.

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 18:47

Commented on [leaspy/io/realizations/collection_realization.py line 57](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-6881891ad3002e05404fa443d2aaba3eR57)



I'm not sure to understand how you've selected the `...=0.01`

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 19:02

Commented on [leaspy/models/utils/attributes/abstract_attributes.py line 60](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..c15af178f041fc6d141515ed4f8c21ffddd93717#diff-d7c16ac3644570cd7f483afcab4b999bR60)



Proposition "Inconsistent attributes: presence of sources for a univariate model"

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 19:09

Commented on [leaspy/models/utils/attributes/abstract_manifold_model_attributes.py line 98](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-abc6c28aa7613989b9a59e3eea4ce1d3L98)



I am not sure why they were present but why did you remove these cheks?

PS : In fact, maybe the previous checks were not strong enough and should list the names of the models that can be used (rather than `X in list[str]`)

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 19:11

Commented on [.gitlab-ci.yml line 2](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-96edf7a6f008de9e928d04e1ae5e12a5L2)



I have no skills on the CI so I'll trust you on that one ;)

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 11, 2021, 19:13

Commented on [leaspy/utils/resampling/general.py line 16](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-c3850c70ed890846909c88f77f190164L16)



Shouldn't we deprecate it soon? I don't know how it should be done in practice, but I think that we should deprecate/remove it asap as there are similar functions in Leaspype

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 10:30

Commented on [leaspy/exceptions.py line 25](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-6d94288d507c946892fa5c04b537c678R25)



This docstring can be found in the online doc here: https://leaspy.readthedocs.io/en/feat_leaspy_exceptions/leaspy.exceptions.html (as well as exception hierarchy).
But I could not manage to align class names like in the docstring (Sphinx / rST is really a pain...)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 11:03

Commented on [tests/functional_tests/api/test_api.py line 177](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-d73622c689a09ffb610b08dd4a2e65dfR177)



Done here: https://leaspy.atlassian.net/wiki/spaces/COL/pages/1080328193/Some+Python+advanced+tricks

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 18:45

Commented on [leaspy/models/utils/__init__.py line 3](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-911a286d4e858cfa5c411c717ffe96e7R3)



Yes indeed I totally agree, currently it is really not proper at all.

We would:
1. store it in `leaspy/models/__init__.py` as you propose
2. or use `leaspy/leaspy/models/data/default_model_settings_*.json` -> these files seem to be totally out-dated and incomplete (I wonder to what extent they are used?!)

In any case, I'd feel uncomfortable having configuration of models at multiple locations, we should have a proper scheme for configuring them.

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 18:52

Commented on [leaspy/io/data/data.py line 160](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-7bf8e54555d1e354ec3979f66cfc3711R160)



Yes we could, actually same applies to `load_cofactors`. I feel that there are many things to refactor in this class: `__next__` (cf. comment -> true Iterator would be needed), `load_cofactors`, this one `to_dataframe`... I didn't dive into that for now...

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 18:55

Commented on [leaspy/algo/fit/abstract_mcmc.py line 74](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-7fc1bee7ec4c0c834f6cbc2c91644f91R74)



Yes I agree...
I kept this nightmare so not to break existing stuff, but I think it'd definitely be a bad practice to let user define a model loss not a model load / definition but later at the fit algorithm step (which is in place currently)!

We should be careful not to break too much code before @PEPoulet merges new models.

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 18:59

Commented on [leaspy/algo/abstract_algo.py line 23](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-23ebe00b397d9ab4e4ababa3805b3079L23)



Everything's here: https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html
(Sphinx and restructuredText are definitely NIGHTMARES)

I don't understand most of errors messages we get when building documentation from sources, it so cryptic... And .rst files are so cryptic as well oO. May be you should consider switching to [MkDocs](https://www.mkdocs.org/) as Clinica does and its extension [MkDocsStrings](https://github.com/mkdocstrings/mkdocstrings) which seem far more modern & convenient

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:01

Commented on [leaspy/io/logs/visualization/plotter.py line 47](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-c9c3894d7799e493bdc4bbb8b3bdefb2R47)



I agree, but we could argue that all logistic models should have axes btw [0, 1]. It was mainly to also cope for `univariate_logistic` models (I think the leaspype equivalent function has this condition written in the same manner)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:20

Commented on [leaspy/algo/abstract_algo.py line 23](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-23ebe00b397d9ab4e4ababa3805b3079L23)



I added the reference in our Sphinx tutorial here: https://leaspy.atlassian.net/wiki/spaces/COL/pages/379093010/Documentation+with+Sphinx+Readthedocs

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:23

Commented on [leaspy/models/abstract_model.py line 16](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-196e847e3c942e778868226a64191c31R16)



Yes I used this flag to avoid circular imports at runtime due to type annotations.
cf. explanation here: https://stackoverflow.com/questions/39740632/python-type-hinting-without-cyclic-imports

When using a static type checker (mypy, or the one shipped with your IDE) TYPE_CHECKING const will be True but in all other cases it will be False (esp. at runtime)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:25

Commented on [leaspy/models/abstract_model.py line 330](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-196e847e3c942e778868226a64191c31L330)



good question can't remember ^^'

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:27

Commented on [CHANGELOG.md line 14](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR14)



Ok we can do this one Wednesday indeed :)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:31

Commented on [leaspy/io/realizations/collection_realization.py line 57](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-6881891ad3002e05404fa443d2aaba3eR57)



I only refactored the existing code in `initialize_from_values` that was completely duplicated in `initialize`, the only difference is this `scale_individual` which was defined with this 0.01 value (I guess in regular case we want true std-dev for individual parameters std, and in "values" cases we want to almost set std-dev to 0, tricked with this factor).

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:32

Commented on [leaspy/utils/subtypes.py line 18](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-3c52fbb29ad1c0ff1fc0510958b73b5dR18)



Indeed, then I'll do this during a Wednesday meeting!

I added functools.wraps in Confluence dev tips I wrote

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:40

Commented on [leaspy/models/utils/attributes/abstract_manifold_model_attributes.py line 98](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-abc6c28aa7613989b9a59e3eea4ce1d3L98)



I removed it because self.name compatibility is already checked previously and in all multivariate cases, velocities is exp'd (cf. original condition) so it was quite pointless!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:41

Commented on [leaspy/utils/resampling/general.py line 16](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-c3850c70ed890846909c88f77f190164L16)



We can remove these deprecated files if you want yes

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:56

Commented on [leaspy/io/logs/visualization/plotting.py line 297](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..c15af178f041fc6d141515ed4f8c21ffddd93717#diff-bcf8dd42601d767f13ba8509eabdf4a1R297)



changed this line in [version 4 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/54/diffs?diff_id=261002659&start_sha=c15af178f041fc6d141515ed4f8c21ffddd93717#50166900ee08ecbea1c4f85927bf7bccfdbe0791_297_297)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:56

Commented on [leaspy/models/utils/attributes/abstract_attributes.py line 60](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..c15af178f041fc6d141515ed4f8c21ffddd93717#diff-d7c16ac3644570cd7f483afcab4b999bR60)



changed this line in [version 4 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/54/diffs?diff_id=261002659&start_sha=c15af178f041fc6d141515ed4f8c21ffddd93717#386659670c7125176d9b7582f10db21d1674d7a0_60_60)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:56

added 1 commit

<ul><li>274050d8 - [DOC/REFACT] minor adjustments for MR54 review</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/54/diffs?diff_id=261002659&start_sha=c15af178f041fc6d141515ed4f8c21ffddd93717)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:56

Commented on [leaspy/io/logs/visualization/plotting.py line 297](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..c15af178f041fc6d141515ed4f8c21ffddd93717#diff-bcf8dd42601d767f13ba8509eabdf4a1R297)



Changed in last commit

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:57

Commented on [leaspy/io/outputs/result.py line 173](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..274050d8b10560b63132edcc7c976d7edeeff8f2#diff-767d0ac58b5e23c2fb2d7c0875ced777R173)



Fixed in last commit

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:57

Commented on [leaspy/io/outputs/result.py line 220](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..274050d8b10560b63132edcc7c976d7edeeff8f2#diff-767d0ac58b5e23c2fb2d7c0875ced777R220)



Fixed in last commit

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:58

Commented on [leaspy/models/abstract_model.py line 330](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-196e847e3c942e778868226a64191c31L330)



I put again a basic type annotation coherent with API function

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:59

Commented on [leaspy/io/outputs/result.py line 173](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..274050d8b10560b63132edcc7c976d7edeeff8f2#diff-767d0ac58b5e23c2fb2d7c0875ced777R173)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/54/diffs?diff_id=261004322&start_sha=274050d8b10560b63132edcc7c976d7edeeff8f2#e3f67df1b2b05af03614f3d34e0c632cc2221226_173_173)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:59

Commented on [leaspy/io/outputs/result.py line 220](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..274050d8b10560b63132edcc7c976d7edeeff8f2#diff-767d0ac58b5e23c2fb2d7c0875ced777R220)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/54/diffs?diff_id=261004322&start_sha=274050d8b10560b63132edcc7c976d7edeeff8f2#e3f67df1b2b05af03614f3d34e0c632cc2221226_220_220)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 13, 2021, 19:59

added 1 commit

<ul><li>3eeade9c - [DOC/REFACT] minor adjustments for MR54 review</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/54/diffs?diff_id=261004322&start_sha=274050d8b10560b63132edcc7c976d7edeeff8f2)

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 19, 2021, 11:29

Commented on [leaspy/models/utils/__init__.py line 3](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-911a286d4e858cfa5c411c717ffe96e7R3)



I agree ! 

Let's keep it as it is and discuss model configuration in !53 (once I rebase it from this MR)

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 19, 2021, 11:46

Commented on [leaspy/io/data/data.py line 160](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-7bf8e54555d1e354ec3979f66cfc3711R160)



Then let's keep this out of this MR !

Should we just merge your comment with the issue #42 to keep track of your remarks once the MR is merged?

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 19, 2021, 11:49

Commented on [leaspy/algo/abstract_algo.py line 23](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-23ebe00b397d9ab4e4ababa3805b3079L23)



Do you feel its worth opening a new issue to switch to MkDocs ? I'm asking you as I don't understand to which extend it is painful, and if the time to switch to MkDocs is worth the time of doing so ? Maybe an 'onboarding' task for Quentin to get a better view of Leaspy ?

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 19, 2021, 11:53

Commented on [leaspy/utils/resampling/general.py line 16](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-c3850c70ed890846909c88f77f190164L16)



Yes, it would be cool if you can do that ! 
We will do that anyway to merge `dev` => `master` (cf #36 )

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 19, 2021, 11:59

Commented on [leaspy/io/logs/visualization/plotter.py line 47](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-c9c3894d7799e493bdc4bbb8b3bdefb2R47)



We could yes, but this means that this "decision" has to be shared and accepted by all. Which is too much effort for such a simple decision.

I'm just afraid that new interns or PhD or devs that write models end up having issues related to somethingout of their hand. But anyway, the plotters are something we will need to refactor at some point !

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 19, 2021, 12:00

@emaheux I've close the threads that seemed okay to me. There are three threads left, help yourself on the way to answer them - or not if not necessary.

Happy to have this merged in `dev` :fire:

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 19, 2021, 12:00

approved this merge request

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 19, 2021, 16:04

Commented on [leaspy/io/data/data.py line 160](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-7bf8e54555d1e354ec3979f66cfc3711R160)



Done!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 19, 2021, 16:37

Commented on [leaspy/algo/abstract_algo.py line 23](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-23ebe00b397d9ab4e4ababa3805b3079L23)



If have really no idea either how time consuming it would be and if it'd really be worth it. I opened an issue as suggested --> #49

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 19, 2021, 16:41

Commented on [leaspy/io/logs/visualization/plotter.py line 47](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-c9c3894d7799e493bdc4bbb8b3bdefb2R47)



Yes I also definitely think that refactoring (/ cleaning) will have to be made on plotters at some point. Let's see by then if there'll be a need to change to this condition!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 19, 2021, 16:45

Commented on [leaspy/utils/resampling/general.py line 16](https://github.com/aramis-lab/leaspy/compare/a89783e17b5cb109fd5bbb809a732eacd08abe5d..3eeade9c7d89d940e58e7b1ec7046010b3d06443#diff-c3850c70ed890846909c88f77f190164L16)



Let's do a separate MR to implement #36 then, to merge this one before it comes too big (which is already a bit the case by my fault ^^')!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Oct 19, 2021, 16:45

mentioned in commit 12d795cbe558bc79cf2edee57adfd900525cb291
